vlasovsolver/gpu_pitch_angle_diffusion.cpp

```diff
@@ -0,0 +1,799 @@
+      CHK_ERR( gpuMalloc((void**)&dev_smallCellIdxArray, maxCellIndex*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_remappedCellIdxArray, numberOfLocalCells*sizeof(size_t)) );
+      CHK_ERR( gpuMalloc((void**)&dev_Ddt, maxCellIndex*sizeof(Real)) );
+      CHK_ERR( gpuMalloc((void**)&dev_potentialDdtValues, maxCellIndex*blocksPerVelocityCell*sizeof(Real)) );
+      CHK_ERR( gpuMalloc((void**)&dev_cellIdxKeys, maxCellIndex*blocksPerVelocityCell*sizeof(int)) );
+
+      // Copy data to device
+      CHK_ERR( gpuMemcpy(dev_cellIdxStartCutoff, host_cellIdxStartCutoff.data(), maxCellIndex*sizeof(size_t), gpuMemcpyHostToDevice) );
+      CHK_ERR( gpuMemcpy(dev_smallCellIdxArray, host_smallCellIdxArray.data(), maxCellIndex*sizeof(size_t), gpuMemcpyHostToDevice) );
+      CHK_ERR( gpuMemcpy(dev_remappedCellIdxArray, host_remappedCellIdxArray.data(), numberOfLocalCells*sizeof(size_t), gpuMemcpyHostToDevice) );
+
+      // Initialize with zero values
+      CHK_ERR( gpuMemset(dev_fmu, 0.0, numberOfLocalCells*nbins_v*nbins_mu*sizeof(Realf)) );
+      CHK_ERR( gpuMemset(dev_fcount, 0, numberOfLocalCells*nbins_v*nbins_mu*sizeof(int)) );
+
+      totalThreadsPerBlock = Hashinator::defaults::MAX_BLOCKSIZE/2; //Using Hashinator::defaults::MAX_BLOCKSIZE/2 = 512 blocks can lead to better streaming multiprocessor occupancy
+      int maxThreadIndex = totalNumberOfVelocityBlocks;
+      blocksPerGrid = (maxThreadIndex+totalThreadsPerBlock-1)/totalThreadsPerBlock;
+
+      // Find cell indices
+      getCellIndexArray_kernel<<<blocksPerGrid, totalThreadsPerBlock>>>(
+         dev_cellIdxArray,
+         dev_velocityIdxArray,
+         dev_cellIdxStartCutoff,
+         totalNumberOfVelocityBlocks,
+         maxCellIndex
+      );
+
+      threadsPerBlock = dim3(WID, WID, WID);
+      blocksPerGrid = maxBlockIndex;
+
+      // Build Fvmu array
+      build2dArrayOfFvmu_kernel<<<blocksPerGrid, threadsPerBlock>>>(
+         dev_cellIdxArray,
+         dev_velocityIdxArray,
+         dev_VBCs,
+         dev_bulkVX,
+         dev_bulkVY,
+         dev_bulkVZ,
+         dev_bValues,
+         dev_fmu,
+         dev_fcount,
+         dVbins,
+         dmubins,
+         nbins_v,
+         nbins_mu
+      );
+      
+      CHK_ERR( gpuPeekAtLastError() );
+      CHK_ERR( gpuDeviceSynchronize() );
+
+      totalThreadsPerBlock = Hashinator::defaults::MAX_BLOCKSIZE/2; //Using Hashinator::defaults::MAX_BLOCKSIZE/2 = 512 blocks can lead to better streaming multiprocessor occupancy
+      maxThreadIndex = numberOfLocalCells*nbins_v*nbins_mu;
+      blocksPerGrid = (maxThreadIndex+totalThreadsPerBlock-1)/totalThreadsPerBlock;
+
+      // Divide by count
+      dividefByCount_kernel<<<blocksPerGrid, totalThreadsPerBlock>>>(
+         dev_smallCellIdxArray,
+         dev_fmu,
+         dev_fcount,
+         nbins_v,
+         nbins_mu,
+         maxThreadIndex
+      );
+      
+      CHK_ERR( gpuPeekAtLastError() );
+      CHK_ERR( gpuDeviceSynchronize() );
```

Comment:
these interim device syncs are also not needed unless you're pulling some values to host